clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java

```diff
@@ -0,0 +1,474 @@
+     */
+    private void unmap(K key) {
+        DynamicKeyMapping<K> mapping = dynamicMapping.orElseThrow(() ->
+            new IllegalStateException("Attempt to unmap key " + key + " which is not dynamically mapped")
+        );
+        fulfillmentMap.remove(key);
+        lookupMap.put(mapping.lookupStrategy.lookupScope(key), key);
+    }
+
+    private void clear(K key) {
+        lookupMap.remove(key);
+        fulfillmentMap.remove(key);
+    }
+
+    OptionalInt keyToBrokerId(K key) {
+        Optional<BrokerScope> scope = fulfillmentMap.getKey(key);
+        if (scope.isPresent()) {
+            return OptionalInt.of(scope.get().destinationBrokerId);
+        } else {
+            return OptionalInt.empty();
+        }
+    }
+
+    /**
+     * Complete the future associated with the given key exceptionally. After is called,
+     * the key will be taken out of both the Lookup and Fulfillment stages so that request
+     * are not retried.
+     */
+    private void completeExceptionally(K key, Throwable t) {
+        KafkaFutureImpl<V> future = futures.get(key);
+        if (future == null) {
+            log.warn("Attempt to complete future for {}, which was not requested", key);
+        } else {
+            clear(key);
+            future.completeExceptionally(t);
+        }
+    }
+
+    /**
+     * Complete the future associated with the given key. After is called, the key will
+     * be taken out of both the Lookup and Fulfillment stages so that request are not retried.
+     */
+    private void complete(K key, V value) {
+        KafkaFutureImpl<V> future = futures.get(key);
+        if (future == null) {
+            log.warn("Attempt to complete future for {}, which was not requested", key);
+        } else {
+            clear(key);
+            future.complete(value);
+        }
+    }
+
+    /**
+     * Check whether any requests need to be sent. This should be called immediately
+     * after the driver is constructed and then again after each request returns
+     * (i.e. after {@link #onFailure(long, RequestSpec, Throwable)} or
+     * {@link #onResponse(long, RequestSpec, AbstractResponse)}).
+     *
+     * @return A list of requests that need to be sent
+     */
+    public List<RequestSpec<K>> poll() {
+        List<RequestSpec<K>> requests = new ArrayList<>();
+        collectLookupRequests(requests);
+        collectFulfillmentRequests(requests);
+        return requests;
+    }
+
+    /**
+     * Get a map of the futures that are awaiting completion.
+     */
+    public Map<K, KafkaFutureImpl<V>> futures() {
+        return futures;
+    }
+
+    /**
+     * Callback that is invoked when a `Call` returns a response successfully.
+     */
+    public void onResponse(
+        long currentTimeMs,
+        RequestSpec<K> spec,
+        AbstractResponse response
+    ) {
+        clearInflightRequest(currentTimeMs, spec);
+        if (spec.scope instanceof AdminApiDriver.BrokerScope) {
```

Comment:
I suggest to add a comment explaining why it is safe to rely on `AdminApiDriver.BrokerScope` to dispatch. At first, I did not understand this until I realized that `AdminApiDriver.BrokerScope` is actually private and can be instantiated only by the `AdminApiDriver`. Do I get this right?

Reply:
Yes, that's right. Maybe I can think of a name which makes this clearer.